infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/fsm/WaitSnapshotApplyState.java

```diff
@@ -139,9 +144,24 @@ public void onEntry(LogReplicationState from) {
             stopSnapshotApply.set(false);
             fsm.getAckReader().markSnapshotSyncInfoOngoing();
         }
+        if (from != this) {
+            snapshotSyncApplyTimerSample = MeterRegistryProvider.getInstance().map(Timer::start);
+        }
         this.fsm.getLogReplicationFSMWorkers().submit(this::verifyStatusOfSnapshotSyncApply);
     }
 
+    @Override
+    public void onExit(LogReplicationState to) {
+        if (to.getType().equals(LogReplicationStateType.IN_LOG_ENTRY_SYNC)) {
+            snapshotSyncApplyTimerSample
+                    .flatMap(sample -> MeterRegistryProvider.getInstance()
+                            .map(registry -> {
+                                Timer timer = registry.timer("logreplication.snapshot.apply.duration");
+                                return sample.stop(timer);
+                            }));
```

Comment:
Yeah I agree it could be closer. But I thought between when message was received and when FSM is under transition (`processEvent -> transition`), we only update some in-memory metadata, the time consumed during those steps should be pretty short? By starting and stopping the timer in `onEntry` and `onExit` I thought we can probably align better to how we define the FSM?

Reply:
Yup, the steps will be very quick, there won't be a significant difference.
Sure, from FSM view point, its cleaner..I was coming from the actual transfer/apply phase view
I'm good with this as well